activation/malfeasance.go

```diff
@@ -0,0 +1,190 @@
+package activation
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/spacemeshos/go-scale"
+	"github.com/spacemeshos/post/shared"
+	"github.com/spacemeshos/post/verifying"
+	"go.uber.org/zap"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/malfeasance/wire"
+	"github.com/spacemeshos/go-spacemesh/signing"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/atxs"
+)
+
+const (
+	multiATXs        = "atx"
+	invalidPostIndex = "invalid_post_index"
+	invalidPrevATX   = "invalid_prev_atx"
+)
+
+type MalfeasanceHandler struct {
+	logger *zap.Logger
+	db     sql.Executor
+
+	edVerifier   *signing.EdVerifier
+	postVerifier PostVerifier
+}
+
+type MalfeasanceOpt func(*MalfeasanceHandler)
+
+func WithMalfeasanceLogger(logger *zap.Logger) MalfeasanceOpt {
+	return func(mh *MalfeasanceHandler) {
+		mh.logger = logger
+	}
+}
+
+func NewMalfeasanceHandler(
+	db sql.Executor,
+	edVerifier *signing.EdVerifier,
+	postVerifier PostVerifier,
+	opt ...MalfeasanceOpt,
+) *MalfeasanceHandler {
+	mh := &MalfeasanceHandler{
+		logger: zap.NewNop(),
+		db:     db,
+
+		edVerifier:   edVerifier,
+		postVerifier: postVerifier,
+	}
+	for _, o := range opt {
+		o(mh)
+	}
+	return mh
+}
+
+func (mh *MalfeasanceHandler) HandleDoublePublish(
+	ctx context.Context,
+	data scale.Type,
+) (types.NodeID, []string, error) {
+	var (
+		firstNid types.NodeID
+		firstMsg wire.AtxProofMsg
+	)
+	ap, ok := data.(*wire.AtxProof)
+	if !ok {
+		return types.EmptyNodeID, []string{multiATXs}, errors.New("wrong message type for multiple ATXs")
+	}
+	for _, msg := range ap.Messages {
+		if !mh.edVerifier.Verify(signing.ATX, msg.SmesherID, msg.SignedBytes(), msg.Signature) {
+			return types.EmptyNodeID, []string{multiATXs}, errors.New("invalid signature")
+		}
+		if firstNid == types.EmptyNodeID {
+			ok, err := atxs.IdentityExists(mh.db, msg.SmesherID)
+			if err != nil {
+				return types.EmptyNodeID, []string{multiATXs},
+					fmt.Errorf("check identity in atx malfeasance %v: %w", msg.SmesherID, err)
+			}
+			if !ok {
+				return types.EmptyNodeID, []string{multiATXs}, fmt.Errorf("identity does not exist: %v", msg.SmesherID)
+			}
+			firstNid = msg.SmesherID
+			firstMsg = msg
+		} else if msg.SmesherID == firstNid {
+			if msg.InnerMsg.PublishEpoch == firstMsg.InnerMsg.PublishEpoch &&
+				msg.InnerMsg.MsgHash != firstMsg.InnerMsg.MsgHash {
+				return msg.SmesherID, []string{multiATXs}, nil
+			}
+		}
+	}
```

Comment:
It seems unnecessarily complicated. We know there are 2 entries only. How about something like that:
```suggestion
	ap, ok := data.(*wire.AtxProof)
	if !ok {
		return types.EmptyNodeID, []string{multiATXs}, errors.New("wrong message type for multiple ATXs")
	}
	for _, msg := range ap.Messages {
		if !mh.edVerifier.Verify(signing.ATX, msg.SmesherID, msg.SignedBytes(), msg.Signature) {
			return types.EmptyNodeID, []string{multiATXs}, errors.New("invalid signature")
		}
	}
	msg0, msg1 := ap.Messages[0], ap.Messages[1]
	ok, err := atxs.IdentityExists(mh.db, msg0.SmesherID)
	if err != nil {
		return types.EmptyNodeID, []string{multiATXs},
			fmt.Errorf("check identity in atx malfeasance %v: %w", msg0.SmesherID, err)
	}
	if !ok {
		return types.EmptyNodeID, []string{multiATXs}, fmt.Errorf("identity does not exist: %v", msg0.SmesherID)
	}

	if msg0.SmesherID == msg1.SmesherID {
			if msg0.InnerMsg.PublishEpoch == msg1.InnerMsg.PublishEpoch &&
				msg0.InnerMsg.MsgHash != msg1.InnerMsg.MsgHash {
				return msg0.SmesherID, []string{multiATXs}, nil
			}
		}
	}
```

Reply:
I just copied the code verbatim from the previous handler and didn't improve on it. I cleaned up the code for this handler a bit in my latest commit